EI-1175 - Add filter(advanced) search

vtermanis opened the pull request:
## Notes
~- Two commits:
    - One illustrates inclusion in existing search service
    - The other shows standalone new service
        - The existing search results format (`SearchResponse`) is still re-used, as is `ResponseType`)~
- Only the synchronous version (a la `SparqlQuery`) is supported initially
- ~`AdvancedSearchRawResponse` is similar to `SparqlQueryResponse`, except that it's more restrictive (fix return contents, i.e. not dynamically defined by user) and thus has its own definition.~ Now out of scope
- Unlike with existing search, there is no separate `expiryTimeout` since that is unused in the old search and timeouts are specified via the headers

## Questions
- Is "Advanced Search" a sensible name to use in the API? ~Set to "_Filter Search_" for now.~
- ~Do both formatted (`SearchResult`) and raw RDF responses need to be supported initially?~ No, only formatted

vtermanis commented on proto/iotics/api/search.proto:
```diff
@@ -33,6 +33,15 @@ service SearchAPI {
   // Receive all search responses associated to a set of Search request for a given client application ID.
   rpc ReceiveAllSearchResponses(SubscriptionHeaders) returns (stream SearchResponse) {}
+
+  // TODO - this will be supported in future but is out of scope for initial requirements
```
Q: Do raw-RDF (as explorer query) and formatted (like classic search) outputs need to be supported immediately?

vtermanis commented on proto/iotics/api/search.proto:
```diff
@@ -83,6 +92,111 @@ message SearchRequest {
+
+// AdvancedSearchRawRequest describes a search request with more filtering possibilities than SearchRequest. It returns
+// encoded RDF about the twins matched by the supplied filter.
+message AdvancedSearchRawRequest {
```
This "duplicate" of `AdvancedSearchRequest` because of a single difference in specifying out format: `ResponseType` VS `RdfResultType`

vtermanis commented on proto/iotics/api/search.proto:
```diff
@@ -83,6 +92,111 @@ message SearchRequest {
+  // Search request payload
+  Payload payload = 3;
+
+  // Search result limit (optional), affecting the maximum number of twins returned, per host. If not specified, no
```
(Limit to be discussed)

vtermanis replied:
@StephanieBracaloni - when we talked about formatted responses and that it should ideally work in a similar manner to the old search, regarding limit/offset we maybe have two options:

1. Limit/Offset like with old search (with maximum/default limit)
    1. Pro: Exactly like before
    2. Pro: Can eventually retrieve all twins
    3. Con: Repeated pagination is made against all hosts and they all have to re-run the query repeatedly even if no more results are available (which could be expensive)
    4. Con: If wanted to move to a multi-result-per-host streaming model, offset would have to be deprecated (which at some point would be a breaking change)
2. Limit only (with maximum/default limit)
    1. Pro: Clients do not have to paginate (and reason about pagination for multiple hosts with various twin counts)
    2. Pro: Hosts with a small number of results do not have to re-run queries for the purpose of pagination (there is none)
    3. Con: There is a limit to how many twins can be returned which could hide some results (but it could be high)
         - Note: In the old search, jena-text matching also has an internal limit (~10k, I believe).
    5. Pro: Either
        1. offset could be introduced later (no breaking change) **or**
        2. multiple responses per host could be enabled (also arguably not breaking, esp. after Conal's update with host-ID present per twin)

What's your view on this? I'd be tempted to go with (2) simply because of point `2.iv` rather than incurring the penalty of `1.iii`. (We can then still choose to add an offset and/or stream multiple responses per host.)
(As long the proto definition supported `Range` in a same way, I assume.)

StephanieBracaloni replied:
I would go with solution 1. (as discussed) so it is in line with the existing behavior and it can be addressed as a whole. 

Plus with 2. all the twins are not reachable when they will be with 1

StephanieBracaloni replied:
> > Nevertheless, at the moment the scope can either be LOCAL or GLOBAL, but what if there was the possibility to search for Twins within a specific remote Host (or a list of) ?
> 
> @Elleppi - that's a good idea (AKA "targeted search") and in fact has also been suggested for other use-cases too.
> 
> This is independent of search filtering capabilities, so we can treat it as a separate feature. If/when this ends up being something that would be very useful for end-users - we can consider it in the goals. (I'm not saying it's not useful already - just that we have to plan it, taking other needs into account.) (CC @StephanieBracaloni)

@Elleppi a good feedback for https://ioticlabs.atlassian.net/wiki/spaces/DT/pages/982384649/Improve+Usability
I will add it :+1:

vtermanis commented:
Extra Qs (from discussion with @conbon ):
1. Would this deprecate the old search (since it could support all of the previous features)
    - Immediate deprecation notice?
3. What should the the RPCs/service be called?

vtermanis commented:
@Elleppi, during our Advanced search discussion on Wednesday you mentioned you had some questions about what the API might look like. Would you mind adding them here to this PR, when you have time?

Elleppi commented:
My only point was whether to have the Advanced Search as part of the Search API or in its own service.
Since most of the times our ext devs will make use of the regular Search, it'd be better in my opinion to keep the Advanced version in its own service

StephanieBracaloni commented on proto/iotics/api/advanced_search.proto:
```diff
@@ -0,0 +1,142 @@
+  // Run an advanced search request on a user timeout and return formatted results. Note that more than one result can
+  // be returned from each host and a final empty response indicates that no more results are expected from the
+  // associated host. (local and remote)
+  rpc AdvancedSearch(AdvancedSearchRequest) returns (stream SearchResponse) {}
```
Re-use an object from another service. Why not add AdvancedSearch (which clearly has a different name) to the existing search service?

StephanieBracaloni replied:
I have seen this comment https://github.com/Iotic-Labs/api/pull/23#issuecomment-1220669866 but I'm not sure to get what is the problem if the name is clearly different

vtermanis replied:
>  Why not add AdvancedSearch (which clearly has a different name) to the existing search service?

The first commit (now c72bc656c7ac1a6457) shows that.
I think the only Q might be: Is "advanced search" the right name for this? (Especially if we end up deprecating the old search later.)

vtermanis replied:
> keep the Advanced version in its own service

@Elleppi , if the new search now basically has almost identical look & feel (with only the request body being different) - maybe having it next to the existing search (with a new name) would also be OK? (In the end it's still "search by twin" - unlike e.g. the meta service which, as you know, is query/update of SPARQL - something different.)

Elleppi replied:
Yes, it would be OK :+1:

vtermanis commented:
Have force-pushed to remove the "raw RDF" rpc

vtermanis commented:
@Elleppi  - I've got one more question for you..

When we spoke to you (and Mihai) about [initial requirements / nice-to-haves](https://ioticlabs.atlassian.net/wiki/spaces/DT/pages/1003126795/Initial+requirements+gathering) for the new search, I remember you mentioning (if I remember correctly) that in particular you thought the following would be useful:
- Ability to compare twin property values using e.g. `>=` or `!=` (i.e. not just equality)
- Ability to compare feed/input property values, not just ones of the twin
- Combine the comparisons using "and"/"or"

From what you know about current/upcoming expected customer API usage, is there anything else that you think is lacking in the "old" search API? (Or is the above a good starting point already?)

Elleppi commented:
@vtermanis all your points are correct and, if all done, they would make a great impact to the usual API usage. This is to say that it is a good starting point already.
Nevertheless, at the moment the `scope` can either be `LOCAL` or `GLOBAL`, but what if there was the possibility to search for Twins within a specific remote Host (or a list of) ?

vtermanis commented:
> Nevertheless, at the moment the scope can either be LOCAL or GLOBAL, but what if there was the possibility to search for Twins within a specific remote Host (or a list of) ?

@Elleppi  - that's a good idea (AKA "targeted search") and in fact has also been suggested for other use-cases too.

This is independent of search filtering capabilities, so we can treat it as a separate feature. If/when this ends up being something that would be very useful for end-users - we can consider it in the goals. (I'm not saying it's not useful already - just that we have to plan it, taking other needs into account.)
(CC @StephanieBracaloni)

vtermanis committed:
EI-1175 - Add Filter Search to Search API

- Has similar behaviour as SynchronousSearch (i.e. streamed
  responses, one per host)

vtermanis commented:
FYI - @Elleppi  & @StephanieBracaloni - how about a slightly different naming:

```
rpc FilterSearch(FilterSearchRequest) returns (stream SearchResponse) {}
```

vtermanis committed:
Makefile - add buf mod update target

Elleppi commented:
IMHO I prefer "AdvancedSearch" as it's slightly more self-explanatory when it is introduced next to the "regular" search, although "FilterSearch" might be more technically correct

vtermanis committed:
fixup! EI-1175 - Add Filter Search to Search API

vtermanis commented:
> IMHO I prefer "AdvancedSearch" as it's slightly more self-explanatory when it is introduced next to the "regular" search, although "FilterSearch" might be more technically correct

@Elleppi  - Done https://github.com/Iotic-Labs/api/pull/23/commits/80cf957329a8e992725bc0cd69d07704a0465019

StephanieBracaloni commented:
> > Nevertheless, at the moment the scope can either be LOCAL or GLOBAL, but what if there was the possibility to search for Twins within a specific remote Host (or a list of) ?
> 
> @Elleppi - that's a good idea (AKA "targeted search") and in fact has also been suggested for other use-cases too.
> 
> This is independent of search filtering capabilities, so we can treat it as a separate feature. If/when this ends up being something that would be very useful for end-users - we can consider it in the goals. (I'm not saying it's not useful already - just that we have to plan it, taking other needs into account.) (CC @StephanieBracaloni)

@Elleppi a good feedback for https://ioticlabs.atlassian.net/wiki/spaces/DT/pages/982384649/Improve+Usability
I will add it :+1:

StephanieBracaloni commented on proto/iotics/api/search.proto:
```diff
@@ -83,6 +86,33 @@ message SearchRequest {
+
+  // Search result limit (optional), affecting the maximum number of twins returned, per host. If not specified, a host-
+  // configured maximum limit will be applied.
+  Limit limit = 4;
```
Offset is mising

vtermanis replied:
As discussed, I'll add offset (or rather: `Range`) in. If we choose to handle pagination/streaming differently in the future then both old & new (advanced) search will be affected in the same way (and we can deprecate/support them in similar fashion).

vtermanis committed:
fixup! EI-1175 - Add Filter Search to Search API

vtermanis commented:
(force pushed rebase against top of `main`)

StephanieBracaloni reviewed (approved):
:+1:
